[popover] Use dialog focus algorithm for dialog popovers

rwlbuis opened the pull request:
#### db3c1dec69db0000cc3f5f3661d86c28bc67e28b
<pre>
[popover] Use dialog focus algorithm for dialog popovers
<a href="https://bugs.webkit.org/show_bug.cgi?id=256208">https://bugs.webkit.org/show_bug.cgi?id=256208</a>

Reviewed by Tim Nguyen.

Implement the spec changes outlines here:
<a href="https://github.com/whatwg/html/pull/9121">https://github.com/whatwg/html/pull/9121</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-dialog-initial-focus-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-dialog-initial-focus.html: Added.
* LayoutTests/platform/ios/fast/dom/focus-dialog-blur-input-type-change-crash-expected.txt:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):

Canonical link: <a href="https://commits.webkit.org/263749@main">https://commits.webkit.org/263749@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ef0a1f9e77757b3f5505ba3f09b654851a8f257

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5747 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7146 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5593 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5594 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5972 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5720 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7357 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5697 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5757 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5013 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7185 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3235 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5032 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12150 "4 flakes 165 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5101 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6919 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5549 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4541 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4963 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1337 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9107 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5363 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1332,6 +1332,11 @@ void HTMLElement::checkAndPossiblyClosePopoverStackInternal()
 static void runPopoverFocusingSteps(HTMLElement& popover)
 {
+    if (is<HTMLDialogElement>(popover)) {
+        downcast<HTMLDialogElement>(popover).runFocusingSteps(true);
```
I commented on the spec PR, but I think running this with `isModal` set to `true` is wrong here.

rwlbuis replied:
I adjusted the patch while waiting for the spec PR issue discussion to be concluded. Do note I adjusted popover-dialog-initial-focus.html to what I think is expected after your suggestion above, hopefully the spec discussion will conclude the same and update the test like that.

nt1m commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -132,18 +136,20 @@ void HTMLDialogElement::queueCancelTask()
 
 // https://html.spec.whatwg.org/multipage/interactive-elements.html#dialog-focusing-steps
-void HTMLDialogElement::runFocusingSteps()
+void HTMLDialogElement::runFocusingSteps(bool isModal)
```
I don't think we need to replace `m_isModal` with this parameter. You can't turn a modal dialog into a popover (`checkPopoverValidity` prevents it), and you can't run `showModal()` on open popovers either.

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -1332,6 +1332,11 @@ void HTMLElement::checkAndPossiblyClosePopoverStackInternal()
 static void runPopoverFocusingSteps(HTMLElement& popover)
 {
+    if (is<HTMLDialogElement>(popover)) {
+        downcast<HTMLDialogElement>(popover).runFocusingSteps(false);
```
```suggestion
    if (auto* dialog = dynamicDowncast<HTMLDialogElement>(popover)) {
        dialog->runFocusingSteps();
```

rwlbuis committed:
[popover] Use dialog focus algorithm for dialog popovers
https://bugs.webkit.org/show_bug.cgi?id=256208

Reviewed by Tim Nguyen.

Implement the spec changes outlines here:
https://github.com/whatwg/html/pull/9121

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-dialog-initial-focus-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-dialog-initial-focus.html: Added.
* LayoutTests/platform/ios/fast/dom/focus-dialog-blur-input-type-change-crash-expected.txt:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):

Canonical link: https://commits.webkit.org/263749@main
